### PR TITLE
memory safety documentation

### PR DESCRIPTION
**Pull request by bdaehlie**
This is based on writing from Alex Gaynor, with his permission.

### CODE REVIEW

File: content/en/docs/memory-safety.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+Under these circumstances, a language that is not memory safe may allow a programmer to read whatever memory contents happen to exist before or after the valid contents of the list. This is called an out of bounds read. The memory before the first item of a list might be the last item of someone else's list. The memory after the last item of a list might be the first item of someone else's list. Accessing this memory would be a severe security vulnerability! Programmers can prevent out of bounds reads by diligently checking the index of the item they're asking for against the length of the list, but programmers make mistakes. It's better to use a memory safe language that protects you and your users from the class of bugs by default.
+
+In a memory safe language we will get an error at compile time or a crash at run time. Crashing the program may seem severe, but it's better than letting users steal each others' data!
+
+A closely related vulnerability is an out-of-bounds write. In this case imagine we tried to change the eleventh or negative first item in our to do list. Now we are changing someone else's to do list!
+
+### Use Afer Free
```
Comment (reaperhulk) [inline content/en/docs/memory-safety.md:24]: Typo here

File: content/en/docs/memory-safety.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+Finally, there is performance. In decades past, one could rely on CPUs getting significantly faster every year or two. This is no longer the case. Instead, CPUs now come with more cores. To take advantage of additional cores, developers are tasked with writing multi-threaded code.
+
+Unfortunately, multi-threading exacerbates the problems associated with a lack of memory safey, As a result, efforts to take advantage of multi-core CPUs are often intractable in C and C++. For example - Mozilla had multiple failed attempts to introduce multi-threading into Firefox's C++ CSS subsystem before finally (successfully) rewriting the system in multi-threaded Rust.
```
Comment (reaperhulk) [inline content/en/docs/memory-safety.md:46]: typo: safety

File: content/en/docs/memory-safety.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+## Is dropping C and C++ really practical?
+
+Hopefully by now we've convinced you that unsafe languages like C and C++ are fundamental root causes of huge swathes of the insecurity in our products, and that while there are practices you can undertake to reduce the risk, you can't get anywhere close to eliminating it. All of which may still leave you with a feeling that changing the programming language you use, to produce millions of lines of code, is an overwhelmingly large task. By breaking it down into manageable pieces, we can start making progress — our goal is not one big-bang rewrite-the-world, but rather to make progress towards reducing our risk.
+
+The first place to start is with brand new projects. For these, you have the choice to simply use a memory safe language. These have the lowest risk, because you do not need to start by rewriting any code, though projects like this often do require improvements to testing or deployment infrastructure to support a new programming language. This was the approach taken in ChromeOS's CrosVM, a brand new component of the operating system.
```
Comment (reaperhulk) [inline content/en/docs/memory-safety.md:74]: Another example possibly worth mentioning: firecracker from AWS.

Comment (bdaehlie) [reply content/en/docs/memory-safety.md:74]: Thanks! I'm going to merge this with the typo fixes then make more editorial changes later.

**Pull request merged**